src/malli/malli.cljc

```diff
@@ -0,0 +1,302 @@
+          (-validator [_ opts]
+            (let [validators (mapv
+                               (fn [[key {:keys [required]} value]]
+                                 (let [valid? (-validator value opts)
+                                       default (not required)]
+                                   (fn [m] (if-let [v (key m)] (valid? v) default))))
+                               entries)]
+              (fn [m] (let [i (.iterator ^Iterable validators)]
+                        (boolean
+                          (loop []
+                            (if (.hasNext i)
+                              (and ((.next i) m) (recur))
+                              true)))))))
+          (-props [_] props)
+          (-form [_] form))))))
+
+(defn registry! [registry]
+  (reduce-kv
+    (fn [acc k v]
+      (reduce
+        (fn [acc alias]
+          (if (contains? acc alias)
+            (fail! ::schema-already-registered {:key k, :alias alias, :registry registry})
+            (assoc acc alias v)))
+        acc (concat [k] (-aliases v))))
+    {} registry))
+
+;;
+;; public api
+;;
+
+(defn schema
+  ([?schema]
+   (schema ?schema nil))
+  ([?schema {:keys [registry] :as opts :or {registry default-registry}}]
+   (if (schema? ?schema)
+     ?schema
+     (let [reg (registry! registry)]
+       (if (vector? ?schema)
+         (apply -into-schema (concat [(get reg (first ?schema))] (props-and-childs (rest ?schema)) [opts]))
+         (if-let [schema' (get reg ?schema)]
+           (-into-schema schema' nil nil opts)
+           (fail! ::invalid-schema {:schema ?schema})))))))
+
+(defn form [?schema]
+  (-form (schema ?schema)))
+
+(defn validator
+  ([?schema]
+   (validator ?schema nil))
+  ([?schema opts]
+   (-validator (schema ?schema opts) opts)))
+
+(defn validate
+  ([?schema value]
+   (validate ?schema value nil))
+  ([?schema value opts]
+   ((validator ?schema opts) value)))
+
+;;
+;; registry
+;;
+
+(def default-registry
+  {:int (-var-schema #'int?)
+   :pos-int (-var-schema #'pos-int?)
+   :neg-int (-var-schema #'neg-int?)
+   :string (-var-schema #'string?)
+   :boolean (-var-schema #'boolean?)
+   :keyword (-var-schema #'keyword?)
+   :and (-composite-schema :and every-pred)
+   :or (-composite-schema :or some-fn)
+   :map (-map-schema)})
+
+;;
+;; helpers
+;;
+
+(def Int (schema :int))
+
+;;
+;; spike
+;;
+
+(ns user)
+
+(require '[malli.malli :as m])
+
+(m/validate m/Int 1)
+
+(m/form [:and int?])
+(m/schema [:and int? pos-int?])
+(m/validate [:and int? [:and pos-int?]] 2)
+(m/validate [:and int? pos-int?] 2)
+
+(m/validate [:and int? [:or pos-int? neg-int?]] 0)
+
+(m/schema [:map [:x int?]])
+
+(m/schema [:map
+           [:x boolean?]
+           [[:opt :y] int?]
+           [:z string?]])
+
+(comment
+  ;; schema-style
+  [:map
+   {:closed true}
+   [:x boolean?]
+   [[:opt :y] int?]
+   [[:req :z] string?]]
+
+  ;; tuples
+  [:map
+   [:x boolean?]
+   [:opt :y int?]
+   [:req :z string?]]
+
+  ;; attrs
+  [:map
```

Comment:
I would vote for this form

Reply:
If key is by default required, is `{:opt true}' only meaningful setting?

Reply:
If `props` => `properties`, I don't the property should be `optional` Instead of `opt`. Let's revisit the closed/open & opt/req later when things look more solid.